Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/checkins/cryptography/CheckInCryptography.kt

```diff
@@ -1,25 +1,135 @@
+            .setLocationIdHash(checkIn.traceLocationIdHash.toProtoByteString())
+            .setMac(mac.toByteString().toProtoByteString())
+            .build()
+    }
+
+    fun decrypt(
+        checkInProtectedReport: CheckInProtectedReport,
+        locationId: ByteArray
+    ): CheckInRecord {
+
+        val macKey = getMacKey(locationId).toByteArray()
+        val mac = getMac(
+            macKey,
+            checkInProtectedReport.iv.toByteArray(),
+            checkInProtectedReport.encryptedCheckInRecord.toByteArray()
+        )
+
+        if (!mac.contentEquals(checkInProtectedReport.mac.toByteArray())) throw IllegalArgumentException(
+            "Message Authentication Codes are not the same ${mac.base64()} != ${
+            checkInProtectedReport.mac.toByteArray().base64()
+            }"
+        )
+
+        val encryptionKey = getEncryptionKey(locationId)
+        val ivParameterSpec = IvParameterSpec(checkInProtectedReport.iv.toByteArray())
+        val decryptedData = aesCryptography.decrypt(
+            encryptionKey.toByteArray(),
+            checkInProtectedReport.encryptedCheckInRecord.toByteArray(),
+            ivParameterSpec
+        )
+
+        return CheckInRecord.parseFrom(decryptedData)
+    }
+
+    private fun getCryptographicSeed(): ByteArray {
+        val cryptographicSeed = ByteArray(16)
+        secureRandom.nextBytes(cryptographicSeed)
+        return cryptographicSeed
+    }
+
+    private fun createCheckInRecord(checkInStart: Instant, checkInEnd: Instant, riskLevel: Int): CheckInRecord {
+        val start = (checkInStart.seconds / 60).toInt()
+        val end = (checkInEnd.seconds / 60).toInt()
+        return CheckInRecord.newBuilder()
+            .setStartIntervalNumber(start)
+            .setPeriod(end - start)
+            .setTransmissionRiskLevel(riskLevel)
+            .build()
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    fun getMac(key: ByteArray, iv: ByteArray, encryptedCheckInRecord: ByteArray): ByteArray {
+        return with(Mac.getInstance(HMAC_SHA256)) {
+            init(SecretKeySpec(key, HMAC_SHA256))
+            doFinal(iv.plus(encryptedCheckInRecord))
+        }
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    fun getEncryptionKey(locationId: ByteArray): ByteString {
+        return CWA_ENCRYPTION_KEY.plus(locationId).toSHA256().decodeHex()
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    fun getMacKey(locationId: ByteArray): ByteString {
+        return CWA_MAC_KEY.plus(locationId).toSHA256().decodeHex()
     }
 
-    fun decrypt(checkInProtectedReport: CheckInProtectedReport): TraceWarning.TraceTimeIntervalWarning {
-        // TODO
-        throw NotImplementedError()
+    companion object {
+        private const val HMAC_SHA256 = "HmacSHA256"
+        private val CWA_MAC_KEY = byteArrayOf(0x43, 0x57, 0x41, 0x2d, 0x4d, 0x41, 0x43, 0x2d, 0x4b, 0x45, 0x59)
```

Comment:
can you please use readable strings as in specs ? it would be easy to find and change ?

Reply:
updated